File: src/host/acp/AgentProcess.ts
import { spawn, type ChildProcessByStdio } from 'node:child_process';
import { createInterface, type Interface } from 'node:readline';
import { Readable, Writable } from 'node:stream';
import * as acp from '@agentclientprotocol/sdk';
import type { AgentDef } from './AgentRegistry';
import { t } from '../i18n';
import { VERSION } from '../version';
import { approveGrokPlan, GROK_EXIT_PLAN, parseGrokExitPlan } from './grokPlan';
import { GROK_ASK_QUESTION, parseGrokQuestion, type GrokQuestionRequest, type GrokQuestionResponse } from './grokQuestions';
import { spawnSpec } from './launch';
import { rewriteExtension } from './subagents/wire';

// What the client side has to accept: updates / permission requests / file reads & writes / questions the agent sends on its own initiative.
// The optional handlers double as capability switches: a handler present is advertised in initialize, an absent one answers method-not-found
export interface ClientHandlers {
  onUpdate: (n: acp.SessionNotification) => void;
  onPermission: (req: acp.RequestPermissionRequest, signal: AbortSignal) => Promise<acp.RequestPermissionResponse>;
  onReadFile?: (req: acp.ReadTextFileRequest) => Promise<acp.ReadTextFileResponse>;
  onWriteFile?: (req: acp.WriteTextFileRequest) => Promise<void>;
  // Form elicitation (elicitation/create with mode=form): the agent asks the user for structured input, e.g. Devin's / Kimi's ask_user_question
  onElicitation?: (req: acp.CreateElicitationRequest, signal: AbortSignal) => Promise<acp.CreateElicitationResponse>;
  // Grok's private question request (`_x.ai/ask_user_question`); without a handler the CLI reports the tool as failed with method-not-found
  onGrokQuestion?: (req: GrokQuestionRequest, signal: AbortSignal) => Promise<GrokQuestionResponse>;
  onStderr?: (line: string) => void;
  onExit?: (code: number | null, signal: NodeJS.Signals | null) => void;
}

export const CLIENT_INFO = { name: 'acpira', version: VERSION };

// The executable would not spawn at all (ENOENT, EACCES, a bad shim): the child 'error' event fires before any
// output. Kept distinct from a process that started and then failed the handshake so health can name the stage
export class AgentSpawnError extends Error {
  constructor(readonly cause: Error) {
    super(cause.message, { cause });
    this.name = 'AgentSpawnError';
  }
}

// A CLI that ignores the polite signal is force-killed after this long (DSH's own graceful-exit window is 5 s)
const KILL_GRACE_MS = 5_000;

// initialize has to answer within this long; a CLI that hangs the handshake would otherwise stall the session forever
const INIT_TIMEOUT_MS = 30_000;

// One agent subprocess = one long-lived ACP connection. stdio carries ndjson; stderr goes line by line to the Output Channel.
// Handlers are rebindable so a warm (initialize-only) process can be handed to a session without a second spawn
export class AgentProcess {
  private constructor(
    readonly def: AgentDef,
    readonly child: ChildProcessByStdio<Writable, Readable, Readable>,
    readonly conn: acp.ClientConnection,
    readonly init: acp.InitializeResponse,
    private readonly stderr: Interface,
    private readonly box: { h: ClientHandlers },
  ) {}

  get agent(): acp.ClientContext { return this.conn.agent; }
  get alive(): boolean { return this.child.exitCode === null && !this.child.killed; }

  bind(h: ClientHandlers) { this.box.h = h; }

  // extraEnv: variables injected by the account layer per identity, layered on top of the agent definition's env.
  // opts.initTimeoutMs bounds the initialize handshake (default INIT_TIMEOUT_MS)
  static async spawn(def: AgentDef, binary: string, cwd: string, h: ClientHandlers, extraEnv?: Record<string, string>, opts?: { initTimeoutMs?: number }): Promise<AgentProcess> {
    const box = { h };
    const spec = spawnSpec(binary, def.args, process.platform, process.env);
    const child = spawn(spec.command, spec.args, {
      cwd,
      env: { ...process.env, ...def.env, ...extraEnv },
      stdio: ['pipe', 'pipe', 'pipe'],
      ...(spec.windowsVerbatimArguments ? { windowsVerbatimArguments: true } : {}),
    });
    const stderr = createInterface({ input: child.stderr });
    stderr.on('line', line => box.h.onStderr?.(line));
    child.on('exit', (code, signal) => box.h.onExit?.(code, signal));

    const raw = acp.ndJsonStream(
      Writable.toWeb(child.stdin) as WritableStream<Uint8Array>,
      Readable.toWeb(child.stdout) as ReadableStream<Uint8Array>,
    );
    // Extension session/update kinds are rewritten into session_info_update before the SDK's closed-union parse drops them
    const stream: acp.Stream = {
      writable: raw.writable,
      readable: raw.readable.pipeThrough(new TransformStream<acp.AnyMessage, acp.AnyMessage>({
        transform(msg, controller) { controller.enqueue(rewriteExtension(msg) as acp.AnyMessage); },
      })),
    };
    const app = acp.client({ name: CLIENT_INFO.name })
      .onNotification(acp.methods.client.session.update, ctx => { box.h.onUpdate(ctx.params); })
      .onRequest(acp.methods.client.session.requestPermission, ctx => box.h.onPermission(ctx.params, ctx.signal))
      .onRequest(GROK_EXIT_PLAN, parseGrokExitPlan, ctx => approveGrokPlan(ctx.params, ctx.signal, box.h.onPermission))
      .onRequest(GROK_ASK_QUESTION, parseGrokQuestion, ctx => {
        if (!box.h.onGrokQuestion) throw acp.RequestError.methodNotFound(GROK_ASK_QUESTION);
        return box.h.onGrokQuestion(ctx.params, ctx.signal);
      })
      .onRequest(acp.methods.client.fs.readTextFile, ctx => {
        if (!box.h.onReadFile) throw acp.RequestError.methodNotFound(acp.methods.client.fs.readTextFile);
        return box.h.onReadFile(ctx.params);
      })
      .onRequest(acp.methods.client.fs.writeTextFile, async ctx => {
        if (!box.h.onWriteFile) throw acp.RequestError.methodNotFound(acp.methods.client.fs.writeTextFile);
        await box.h.onWriteFile(ctx.params);
        return {};
      })
      .onRequest(acp.methods.client.elicitation.create, ctx => {
        if (!box.h.onElicitation) throw acp.RequestError.methodNotFound(acp.methods.client.elicitation.create);
        return box.h.onElicitation(ctx.params, ctx.signal);
      });
    const conn = app.connect(stream);

    const exited = new Promise<never>((_, reject) => {
      child.once('exit', (code, signal) => reject(new Error(t('host.spawnExited', { command: def.command, code: code ?? '-', signal: signal ?? '-' }))));
      child.once('error', e => reject(new AgentSpawnError(e)));
    });
    const initReq = {
      protocolVersion: acp.PROTOCOL_VERSION,
      clientInfo: CLIENT_INFO,
      clientCapabilities: {
        fs: { readTextFile: !!h.onReadFile, writeTextFile: !!h.onWriteFile },
        terminal: false,
        // The host can reproduce the agent's invocation in an interactive terminal, so `type: 'terminal'`
        // authMethods may be offered (claude-agent-acp only advertises its logins to clients that declare this);
        // an agent whose ACP process ignores the local login opts out (AgentDef.auth.terminal, Devin)
        ...(def.auth?.terminal === false ? {} : { auth: { terminal: true } }),
        ...(h.onElicitation ? { elicitation: { form: {} } } : {}),
        // ACP boolean session config options (RFD boolean-config-option); the codex / claude adapters degrade
        // an option like Codex's fast-mode to an on/off select when the client does not declare this
        session: { configOptions: { boolean: {} } },
        ...(def.subagents === false ? {} : { subagents: {} }),
        _meta: {
          // codex-acp streams shell output as _meta.terminal_output_delta and only then drops its JSON rawOutput receipt;
          // claude-agent-acp switches Bash to an agent-managed terminal on either terminal_output flag
          terminal_output_delta: true,
          // JetBrains AIR bridge (RFD #1992 + adapter extensions): nativeSubagentSessions follows the def's subagents
          // gate; sessionFailure and asyncTasks are independent capabilities the host always supports.
          // recommendedValue: claude-agent-acp drops its placeholder `default` effort row (and the `default` model row when
          // a concrete model matches it) and reports concrete current values instead
          jetbrains: { air: { version: 1, capabilities: [...(def.subagents === false ? [] : ['nativeSubagentSessions']), 'sessionFailure', 'asyncTasks', 'recommendedValue'] } },
        },
      },
    } as acp.InitializeRequest;
    const initTimeoutMs = opts?.initTimeoutMs ?? INIT_TIMEOUT_MS;
    let initTimer: ReturnType<typeof setTimeout> | undefined;
    const timedOut = new Promise<never>((_, reject) => {
      initTimer = setTimeout(() => reject(new Error(t('host.initTimeout', { command: def.command, seconds: initTimeoutMs / 1000 }))), initTimeoutMs);
    });
    // A CLI that answers initialize with an error is still running; without this it would sit there as an orphan behind the error notice
    try {
      const init: acp.InitializeResponse = await Promise.race([conn.agent.request(acp.methods.agent.initialize, initReq), exited, timedOut]);
      return new AgentProcess(def, child, conn, init, stderr, box);
    } catch (e) {
      stderr.close();
      conn.close();
      terminate(child);
      throw e;
    } finally {
      clearTimeout(initTimer);
    }
  }

  kill() {
    const exited = this.child.exitCode !== null || this.child.signalCode !== null
      ? Promise.resolve()
      : new Promise<void>(resolve => { this.child.once('exit', () => resolve()); });
    this.stderr.close();
    this.conn.close();
    terminate(this.child);
    return exited;
  }
}

function terminate(child: ChildProcessByStdio<Writable, Readable, Readable>) {
  if (child.exitCode !== null || child.killed) return;
  child.kill();
  const force = setTimeout(() => { if (child.exitCode === null) child.kill('SIGKILL'); }, KILL_GRACE_MS);
  force.unref();
  child.once('exit', () => clearTimeout(force));
}

File: rust/crates/acpira-host/src/acp/agent_process.rs
//! One agent subprocess = one long-lived ACP connection. Handlers are
//! rebindable so a warm (initialize-only) process can be handed to a session without a second spawn

use std::process::Stdio;
use std::sync::Arc;
use std::time::Duration;

use anyhow::{Result, anyhow};
use serde_json::{Value, json};
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use tokio::sync::watch;

use super::agent_registry::AgentDef;
use super::cancel::Cancel;
use super::grok::{GROK_ASK_QUESTION, GROK_EXIT_PLAN};
use super::launch::{Os, ProcessEnv, spawn_spec};
use super::rpc::{BoxFuture, Connection, Inbound, RpcError};
use crate::i18n::tp;

pub const CLIENT_NAME: &str = "acpira";
pub const PROTOCOL_VERSION: i64 = 1;
const KILL_GRACE: Duration = Duration::from_secs(5);
pub const INIT_TIMEOUT: Duration = Duration::from_secs(30);

pub fn client_version() -> &'static str {
  env!("CARGO_PKG_VERSION")
}

/// What the client side accepts from the agent. Every handler here is advertised; the session and the idle stubs both
/// implement all of them
pub trait ClientHandlers: Send + Sync + 'static {
  fn on_update(&self, params: Value);
  fn on_permission(&self, req: Value, cancel: Cancel) -> BoxFuture<Result<Value, RpcError>>;
  fn on_elicitation(&self, req: Value, cancel: Cancel) -> BoxFuture<Result<Value, RpcError>>;
  fn on_grok_question(&self, req: Value, cancel: Cancel) -> BoxFuture<Result<Value, RpcError>>;
  fn on_stderr(&self, _line: &str) {}
  fn on_exit(&self, _code: Option<i32>, _signal: Option<String>) {}
}

/// The executable would not spawn at all (ENOENT, EACCES): distinct from a process that started and then failed the handshake
#[derive(Debug)]
pub struct AgentSpawnError(pub std::io::Error);

impl std::fmt::Display for AgentSpawnError {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    let code = match self.0.kind() {
      std::io::ErrorKind::NotFound => "ENOENT",
      std::io::ErrorKind::PermissionDenied => "EACCES",
      _ => "",
    };
    if code.is_empty() { write!(f, "spawn failed: {}", self.0) } else { write!(f, "spawn {code}") }
  }
}

impl std::error::Error for AgentSpawnError {}

type HandlerBox = Arc<parking_lot::RwLock<Arc<dyn ClientHandlers>>>;

struct Router {
  box_: HandlerBox,
}

impl Inbound for Router {
  fn notification(&self, method: &str, params: Value) {
    if method == "session/update" {
      self.box_.read().clone().on_update(params);
    }
  }

  fn request(&self, method: String, params: Value, cancel: Cancel) -> BoxFuture<Result<Value, RpcError>> {
    let h = self.box_.read().clone();
    match method.as_str() {
      "session/request_permission" => h.on_permission(params, cancel),
      "elicitation/create" => h.on_elicitation(params, cancel),
      GROK_ASK_QUESTION => match super::grok::parse_question(&params) {
        Ok(req) => h.on_grok_question(req, cancel),
        Err(e) => Box::pin(async move { Err(e) }),
      },
      GROK_EXIT_PLAN => match super::grok::parse_exit_plan(&params) {
        Ok(req) => Box::pin(super::grok::approve_plan(req, cancel, h)),
        Err(e) => Box::pin(async move { Err(e) }),
      },
      _ => Box::pin(async move { Err(RpcError::method_not_found(&method)) }),
    }
  }
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
enum Life {
  Running,
  Exited,
}

pub struct AgentProcess {
  pub def: AgentDef,
  pub conn: Connection,
  /// The initialize response as the agent sent it
  pub init: Value,
  pub pid: Option<u32>,
  box_: HandlerBox,
  life: watch::Receiver<Life>,
  killed: parking_lot::Mutex<bool>,
}

impl AgentProcess {
  pub fn alive(&self) -> bool {
    *self.life.borrow() == Life::Running && !*self.killed.lock()
  }

  pub fn bind(&self, h: Arc<dyn ClientHandlers>) {
    *self.box_.write() = h;
  }

  pub async fn request(&self, method: &str, params: Value) -> Result<Value, RpcError> {
    self.conn.request(method, params).await
  }

  pub async fn request_ordered(&self, method: &str, params: Value) -> Result<(Value, super::rpc::Handoff), RpcError> {
    self.conn.request_ordered(method, params).await
  }

  pub fn notify(&self, method: &str, params: Value) {
    self.conn.notify(method, params);
  }

  /// agentCapabilities as advertised (Null when absent)
  pub fn caps(&self) -> &Value {
    self.init.get("agentCapabilities").unwrap_or(&Value::Null)
  }

  /// `extra_env`: variables the account layer injects per identity, on top of the definition's env
  pub async fn spawn(
    def: &AgentDef,
    binary: &str,
    cwd: &str,
    h: Arc<dyn ClientHandlers>,
    extra_env: Option<&acpira_shared::transcript::StrMap>,
    init_timeout: Option<Duration>,
  ) -> Result<Arc<AgentProcess>> {
    let box_: HandlerBox = Arc::new(parking_lot::RwLock::new(h));
    let spec = spawn_spec(binary, &def.args, Os::current(), &ProcessEnv);
    let mut cmd = Command::new(&spec.command);
    #[cfg(windows)]
    if spec.verbatim {
      for a in &spec.args {
        cmd.raw_arg(a);
      }
    } else {
      cmd.args(&spec.args);
    }
    #[cfg(not(windows))]
    cmd.args(&spec.args);
    cmd.current_dir(cwd).stdin(Stdio::piped()).stdout(Stdio::piped()).stderr(Stdio::piped()).kill_on_drop(false);
    for (k, v) in def.env.iter().flatten() {
      cmd.env(k, v);
    }
    for (k, v) in extra_env.into_iter().flatten() {
      cmd.env(k, v);
    }
    let mut child = cmd.spawn().map_err(|e| anyhow::Error::new(AgentSpawnError(e)))?;
    let pid = child.id();
    let stdin = child.stdin.take().expect("piped");
    let stdout = child.stdout.take().expect("piped");
    let stderr = child.stderr.take().expect("piped");

    let err_box = box_.clone();
    tokio::spawn(async move {
      let mut lines = BufReader::new(stderr).lines();
      while let Ok(Some(line)) = lines.next_line().await {
        err_box.read().clone().on_stderr(&line);
      }
    });

    let router_box = box_.clone();
    let inbound: Arc<dyn Fn() -> Arc<dyn Inbound> + Send + Sync> = {
      let router: Arc<dyn Inbound> = Arc::new(Router { box_: router_box });
      Arc::new(move || router.clone())
    };
    let log_box = box_.clone();
    let conn = Connection::start(stdout, stdin, inbound, Arc::new(move |l: &str| log_box.read().clone().on_stderr(l)));

    let (life_tx, life_rx) = watch::channel(Life::Running);
    let exit_box = box_.clone();
    let exit_conn = conn.clone();
    let (exit_info_tx, exit_info_rx) = watch::channel::<Option<(Option<i32>, Option<String>)>>(None);
    tokio::spawn(async move {
      let status = child.wait().await;
      let (code, signal) = match status {
        Ok(s) => (s.code(), signal_name(&s)),
        Err(_) => (None, None),
      };
      let _ = exit_info_tx.send(Some((code, signal.clone())));
      let _ = life_tx.send(Life::Exited);
      exit_conn.close();
      exit_box.read().clone().on_exit(code, signal);
    });

    let init_req = initialize_request(def);
    let timeout = init_timeout.unwrap_or(INIT_TIMEOUT);
    let mut exit_rx = exit_info_rx;
    let outcome = tokio::select! {
      r = conn.request("initialize", init_req) => r.map_err(anyhow::Error::new),
      seen = exit_rx.wait_for(|v| v.is_some()) => {
        let (code, signal) = seen.ok().and_then(|v| v.clone()).unwrap_or((None, None));
        let code = code.map(|c| c.to_string()).unwrap_or_else(|| "-".into());
        Err(anyhow!(tp("host.spawnExited", &[("command", &def.command), ("code", &code), ("signal", signal.as_deref().unwrap_or("-"))])))
      }
      _ = tokio::time::sleep(timeout) => Err(anyhow!(tp("host.initTimeout", &[("command", &def.command), ("seconds", &timeout.as_secs().to_string())]))),
    };
    match outcome {
      Ok(init) => {
        Ok(Arc::new(AgentProcess { def: def.clone(), conn, init, pid, box_, life: life_rx, killed: parking_lot::Mutex::new(false) }))
      }
      Err(e) => {
        // A CLI that answered initialize with an error is still running: never leave it behind as an orphan
        conn.close();
        terminate(pid, life_rx);
        Err(e)
      }
    }
  }

  fn kill_now(&self) {
    *self.killed.lock() = true;
    self.conn.close();
    terminate(self.pid, self.life.clone());
  }

  /// SIGKILL at once, for a host that is about to exit and cannot wait out the grace period
  pub fn kill_hard(&self) {
    *self.killed.lock() = true;
    self.conn.close();
    if let Some(pid) = self.pid
      && *self.life.borrow() != Life::Exited
    {
      send_signal(pid, true);
    }
  }

  /// SIGTERM, then SIGKILL after the grace period; resolves when the process has exited
  pub async fn kill(&self) {
    self.kill_now();
    let mut life = self.life.clone();
    let _ = life.wait_for(|l| *l == Life::Exited).await;
  }
}

fn terminate(pid: Option<u32>, life: watch::Receiver<Life>) {
  if *life.borrow() == Life::Exited {
    return;
  }
  let Some(pid) = pid else { return };
  send_signal(pid, false);
  let mut life = life;
  tokio::spawn(async move {
    if tokio::time::timeout(KILL_GRACE, life.wait_for(|l| *l == Life::Exited)).await.is_err() {
      send_signal(pid, true);
    }
  });
}

#[cfg(unix)]
fn send_signal(pid: u32, force: bool) {
  // SAFETY: plain kill(2) on our own child's pid
  unsafe {
    libc::kill(pid as libc::pid_t, if force { libc::SIGKILL } else { libc::SIGTERM });
  }
}

#[cfg(not(unix))]
fn send_signal(pid: u32, _force: bool) {
  let _ =
    std::process::Command::new("taskkill").args(["/PID", &pid.to_string(), "/T", "/F"]).stdout(Stdio::null()).stderr(Stdio::null()).spawn();
}

#[cfg(unix)]
fn signal_name(s: &std::process::ExitStatus) -> Option<String> {
  use std::os::unix::process::ExitStatusExt;
  let sig = s.signal()?;
  Some(
    match sig {
      libc::SIGTERM => "SIGTERM",
      libc::SIGKILL => "SIGKILL",
      libc::SIGINT => "SIGINT",
      libc::SIGHUP => "SIGHUP",
      libc::SIGSEGV => "SIGSEGV",
      libc::SIGABRT => "SIGABRT",
      libc::SIGPIPE => "SIGPIPE",
      _ => return Some(format!("SIG{sig}")),
    }
    .to_owned(),
  )
}

#[cfg(not(unix))]
fn signal_name(_: &std::process::ExitStatus) -> Option<String> {
  None
}

/// The initialize request, capability by capability as the TS host advertises it
pub fn initialize_request(def: &AgentDef) -> Value {
  let mut caps = json!({
    "fs": { "readTextFile": false, "writeTextFile": false },
    "terminal": false,
  });
  // The host can reproduce the agent's invocation in an interactive terminal; an agent whose ACP process ignores the local login opts out
  if def.terminal_auth {
    caps["auth"] = json!({ "terminal": true });
  }
  caps["elicitation"] = json!({ "form": {} });
  // ACP boolean session config options (RFD boolean-config-option). `compaction` opts into structured
  // compaction_update: without it claude-agent-acp 0.81.0 reports compaction as a `think` tool call
  caps["session"] = json!({ "configOptions": { "boolean": {} }, "compaction": {} });
  if def.subagents {
    caps["subagents"] = json!({});
  }
  let mut air = vec![];
  if def.subagents {
    air.push("nativeSubagentSessions");
  }
  air.extend(["sessionFailure", "asyncTasks", "recommendedValue"]);
  caps["_meta"] = json!({
    "terminal_output_delta": true,
    "jetbrains": { "air": { "version": 1, "capabilities": air } },
  });
  json!({ "protocolVersion": PROTOCOL_VERSION, "clientInfo": { "name": CLIENT_NAME, "version": client_version() }, "clientCapabilities": caps })
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn initialize_advertises_structured_compaction() {
    let req = initialize_request(&AgentDef::default());
    assert!(req["clientCapabilities"]["session"]["compaction"].is_object());
  }
}
